### PR TITLE
Bin Movements For Wip/Failed MRL bins

### PR DESCRIPTION
**Pull request by lmatoti**
Bin Movements For Wip/Failed MRL bins
- Added columns 'wip_disallowed' and 'block_mrl_failures' to locations table
- Change locations CRUD to allow setting and unsetting of these two flags
- Changed all putaway apps to disallow the move if bin(s) are in WIP AND if destination location.wip_disallowed is true
- Changed all putaway apps to disallow the move if bin(s) fail mrl results AND location.block_mrl_failures is true

### CODE REVIEW

File: lib/finished_goods/services/move_stock.rb [hunk 1]
```diff
@@ -135,6 +135,32 @@ def validate
...
+      end
+      success_response('ok')
+    end
+
+    def validate_block_move_for_wip?
+      if stock_type == AppConst::BIN_STOCK_TYPE
+        wip_disallowed = repo.get_value(:locations, :wip_disallowed, id: location_to_id)
+        bin_asset_number = repo.get_value(:rmt_bins, :bin_asset_number, id: stock_item_id)
+        res = repo.check_bin_in_wip(bin_asset_number)
+        return failed_response("This location disallows WIP (#{res.instance}) bins") if !res.success && wip_disallowed
```
Comment (jcsjcs) [inline lib/finished_goods/services/move_stock.rb:162]: You already have the bin id, so simpler, more efficient code would be:
```ruby
wip_bin = repo.select_values(:wip_bins, :id, rmt_bin_id: stock_item_id).length.positive?
```

**Review by jcsjcs — CHANGES_REQUESTED** [review]

File: lib/finished_goods/services/move_stock.rb [hunk 2]
```diff
@@ -135,11 +135,43 @@ def validate
       res = validate_is_to_location_sealed
       return res unless res.success
 
+      res = validate_block_move_for_wip?
```
Comment (jcsjcs) [inline lib/finished_goods/services/move_stock.rb:138]: Currently `res` will be nil for a pallet.

File: lib/finished_goods/services/move_stock.rb [hunk 3]
```diff
@@ -135,11 +135,43 @@ def validate
...
+      return res unless res.success
+
+      res = block_move_for_mrl_failures?
```
Comment (jcsjcs) [inline lib/finished_goods/services/move_stock.rb:141]: Currently `res` will be nil for a pallet.

File: lib/finished_goods/services/move_stock.rb [hunk 4]
```diff
@@ -135,11 +135,43 @@ def validate
...
+        return failed_response('This location disallows bins that fail MRL') if res[:errors][:failed]
+      end
+      success_response('ok')
+    end
+
+    def validate_block_move_for_wip?
+      if stock_type == AppConst::BIN_STOCK_TYPE
+        return ok_response unless @wip_disallowed
+
+        wip_bin_id, context = repo.get_value(:wip_bins, %i[id context], rmt_bin_id: stock_item_id)
```
Comment (jcsjcs) [inline lib/finished_goods/services/move_stock.rb:162]: There can be more than one record for the same bin in `wip_bins`.

File: lib/messcada/repositories/messcada_repo.rb [hunk 5]
```diff
@@ -706,7 +706,7 @@ def check_bin_in_wip(bin_number)
 
       return ok_response unless context
 
-      failed_response("Bin in WIP: #{context}")
+      failed_response("Bin in WIP: #{context}", context)
```
Comment (jcsjcs) [inline lib/messcada/repositories/messcada_repo.rb:709]: This change is no longer required.

**Review by jcsjcs — CHANGES_REQUESTED** [review]

File: lib/finished_goods/services/move_stock.rb (hunk 3 above)
Comment (lmatoti) [reply lib/finished_goods/services/move_stock.rb:141]: For a pallet res will be 'success_response('ok')'

File: lib/finished_goods/services/move_stock.rb (hunk 2 above)
Comment (lmatoti) [reply lib/finished_goods/services/move_stock.rb:138]: Same here, res will be 'success_response('ok')' for a pallet

**Pull request merged**